frontend/src/components/main-page/utils/service-docs-utils.ts

```diff
@@ -0,0 +1,104 @@
+import { GetServiceDocResponse } from 'msadoc-client';
+
+export interface ServiceDocsGroup {
+  /**
+   * The `name` of a group is the last part of its `identifier`.
+   *
+   * Example: If the `identifier` is `foo.bar.baz`, then the `name` is `baz`.
+   */
+  name: string;
+  /**
+   * The unique identifier of this group like `foo.bar.baz`.
+   */
+  identifier: string;
+
+  /**
+   * The direct children of this group.
+   *
+   * The object keys are the names of the respective child group.
+   * (Compared to just using an array, this enables a faster lookup of groups.)
+   */
+  childGroups: { [groupName: string]: ServiceDocsGroup };
+
+  /**
+   * The services that belong to this group.
+   */
+  services: GetServiceDocResponse[];
+}
+export type ServiceDocsRootGroup = Omit<
+  ServiceDocsGroup,
+  'name' | 'identifier'
+>;
+
+export function buildGroupsTree(
+  serviceDocs: GetServiceDocResponse[],
+): ServiceDocsRootGroup {
+  const rootGroup: ServiceDocsRootGroup = {
+    childGroups: {},
+    services: [],
+  };
+
+  for (const singleServiceDoc of serviceDocs) {
+    let group = rootGroup;
+
+    // By default, we add the service to the root group. But if the service has the `group` attribute, then we want to add it to this specified group instead.
+    if (singleServiceDoc.group !== undefined) {
+      group = addGroupIfNotExists(singleServiceDoc.group, rootGroup);
+    }
+
+    group.services.push(singleServiceDoc);
+  }
+
+  return rootGroup;
+}
+
+export function getGroupByIdentifier(
+  groupIdentifier: string,
+  groupsTree: ServiceDocsRootGroup,
+): ServiceDocsGroup | undefined {
+  const splitGroupIdentifier = groupIdentifier.split('.');
+  let currentGroup = groupsTree;
+  for (const identifierPart of splitGroupIdentifier) {
+    const nextGroup = currentGroup.childGroups[identifierPart];
+    if (!nextGroup) {
+      return undefined;
+    }
+    currentGroup = nextGroup;
+  }
+  return currentGroup as ServiceDocsGroup;
+}
+
+function addGroupIfNotExists(
+  groupIdentifier: string,
+  rootGroup: ServiceDocsRootGroup,
+): ServiceDocsGroup | ServiceDocsRootGroup {
+  const splitGroupIdentifier = groupIdentifier.split('.');
+  if (splitGroupIdentifier[0] === undefined) {
+    // This should not happen.
```

Comment:
Should wr log sth in these cases? Also see cases below.

Reply:
Hmmm this line only exists because TS does not know that `someString.split()` always returns at least one element.(*) So this point should in reality never be reached. But in the future, someone might refactor this and introduce an actual bug. So IMHO you are right - I think we should actually log something here. I would suggest to go the really easy way and to just log "This should not happen." or "This point should not be reached.". I mean - this is really just to tell the developer that he/she messed something up really badly, so I assume that such an error message should be sufficient (the developer can/should then use their devtools to determine what really caused the issue). This makes our life easier because otherwise, we will have to invent an error message for all such cases. What do you think?

(*) FYI: AFAIK, there is only one special case where `someString.split()` reuturns an empty array: That happens if both the string being split, and the argument passed to the function, are the emptystring, i.e. `"".split("")` returns `[]`. That's probably(?) the reason why TS behaves like this.